Migrate sidebar item pref for built-in items which do not specify built-in-item-type

petemill opened the pull request:
Avoid crash by not trying to load them, but also migrate these items to include the item type key, bringing back the function to find built-in item type from url.

This crash was occuring on Nightly for some users who had modified sidebar preferences close to the time that it was initially released and not modified the preferences since.

<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves https://github.com/brave/brave-browser/issues/24965
Additionally, from https://github.com/brave/brave-core/pull/14800:
Resolves https://github.com/brave/internal/issues/899
Resolves https://github.com/brave/internal/issues/900
Resolves https://github.com/brave/internal/issues/901

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
On any profile, open the Preferences file in a code editor, and replace the "brave.sidebar" section with the following:
```
    "sidebar": {
      "sidebar_items": [
        {
          "open_in_panel": true,
          "title": "Brave Together",
          "type": 0,
          "url": "https://together.brave.com/"
        },
        {
          "open_in_panel": false,
          "title": "Crypto Wallets",
          "type": 0,
          "url": "chrome://wallet/"
        },
        {
          "open_in_panel": true,
          "title": "Bookmarks",
          "type": 0,
          "url": "chrome://bookmarks/"
        },
        {
          "open_in_panel": true,
          "title": "History",
          "type": 0,
          "url": "chrome://history/"
        },
        {
          "open_in_panel": false,
          "title": "Settings - About Brave",
          "type": 1,
          "url": "chrome://settings/help"
        }
      ]
    },
```

petemill commented on components/sidebar/sidebar_service.cc:
```diff
@@ -335,7 +407,15 @@ std::vector<SidebarItem> SidebarService::GetHiddenDefaultSidebarItems() const {
                  added_item.built_in_item_type;
         });
-    default_items.erase(iter);
+    DCHECK(iter != default_items.end());
+    if (iter == default_items.end()) {
+      // This should not happen, since |LoadSidebarItems| ignores items which
+      // aren't included in |GetDefaultSidebarItems|.
+      LOG(ERROR) << "Bad built-in item loaded, cannot match to : "
+                 << static_cast<int>(added_item.type);
+    } else {
+      default_items.erase(iter);
+    }
```
This is where the crash was. It shouldn't get here anymore because of the fix earlier in the stack, but it's better to keep a check here.

petemill commented on components/sidebar/sidebar_service.cc:
```diff
@@ -393,23 +473,27 @@ void SidebarService::LoadSidebarItems() {
-          items_.emplace_back(*std::make_move_iterator(iter));
-          default_items_to_add.erase(iter);
+        const auto value = item.FindIntKey(kSidebarItemBuiltInItemTypeKey);
+        if (!value.has_value()) {
```
This line solves the issue since we should skip parsing the item if we have bad data. The previous version of this line was missing the `continue`. Plus it isn't immediately obvious that:
```
if (const auto value =
                item.FindIntKey(kSidebarItemBuiltInItemTypeKey)
```
is the same as:
```
        const auto value = item.FindIntKey(kSidebarItemBuiltInItemTypeKey);
        if (!value.has_value()) 
```

petemill commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -421,6 +421,64 @@ TEST_F(SidebarServiceTest, MigratePrefSidebarBuiltInItemsNoneHidden) {
 
+// Verify service has migrated the previous pref format where built-in items
+// had url stored and not built-in-item-type.
+TEST_F(SidebarServiceTest, MigratePrefSidebarBuiltInItemsNoType) {
```
I verified that this new test fails with the version of SidebarService previous to this PR, and passes with this PR.

petemill commented on components/sidebar/sidebar_service.cc:
```diff
@@ -393,23 +473,29 @@ void SidebarService::LoadSidebarItems() {
-          default_items_to_add.erase(iter);
+        const auto built_in_type_value =
+            item.FindIntKey(kSidebarItemBuiltInItemTypeKey);
+        if (!built_in_type_value.has_value()) {
+          VLOG(1) << "built-in item did not have a type: "
+                  << item.DebugString();
           continue;
```
These lines solve the issue since we should skip parsing the item if we have bad data. The previous version of this line was missing the `continue`. Plus it isn't immediately obvious that:
```
if (const auto value =
                item.FindIntKey(kSidebarItemBuiltInItemTypeKey)
```
is the same as:
```
        const auto value = item.FindIntKey(kSidebarItemBuiltInItemTypeKey);
        if (!value.has_value()) 
```

cdesouza-chromium commented on components/sidebar/sidebar_service.cc:
```diff
@@ -335,7 +407,15 @@ std::vector<SidebarItem> SidebarService::GetHiddenDefaultSidebarItems() const {
                  added_item.built_in_item_type;
         });
-    default_items.erase(iter);
+    DCHECK(iter != default_items.end());
```
It is against the chromium style guide to have a DCHECK and a check just after. Please remove this check, or the conditional if after.

simonhong replied:
I also think using this DCHECK alone would be fine.

cdesouza-chromium replied:
Notice that I have a different PR that makes these changes here unnecessary, and I plan to merge it before this PR.

petemill replied:
I have rebased this PR on your PR now @cdesouza-chromium. Pushing in a minute

cdesouza-chromium commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -421,6 +421,64 @@ TEST_F(SidebarServiceTest, MigratePrefSidebarBuiltInItemsNoneHidden) {
+        "chrome://history/",
+    };
+    for (const auto& url : urls) {
+      base::Value dict(base::Value::Type::DICTIONARY);
```
Please use base::Value::Dict here as all dictionary interfaces in base::Value are deprecated, otherwise this is going to be revisited for refactoring.

petemill replied:
I'd prefer to convert all the instances of it in this file (9 of them) as a follow-up to keep this PR minimal, but also because it doesn't seem quickly clear to me how to give a `base::Value::Dict` to `ListPrefUpdate::Append`

petemill replied:
Actually I found out it's simply `base::Value(dict)`

cdesouza-chromium replied:
Awesome

cdesouza-chromium commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -421,6 +421,64 @@ TEST_F(SidebarServiceTest, MigratePrefSidebarBuiltInItemsNoneHidden) {
+      update->Append(std::move(dict));
+    }
+    // Add a custom item to make sure we don't interfere with it
+    base::Value dict(base::Value::Type::DICTIONARY);
```
base::Value::Dict

cdesouza-chromium commented on components/sidebar/sidebar_service_unittest.cc:
```diff
@@ -421,6 +421,64 @@ TEST_F(SidebarServiceTest, MigratePrefSidebarBuiltInItemsNoneHidden) {
+  InitService();
+
+  // Verify migration
+  auto* preference = prefs_.FindPreference(kSidebarItems);
```
Consider using `GetValueList` here

cdesouza-chromium commented on components/sidebar/sidebar_service.cc:
```diff
@@ -196,11 +225,41 @@ void SidebarService::MigratePrefSidebarBuiltInItemsToHidden() {
+        items_are_modified = true;
+        continue;
+      }
+      const auto item_id_parsed = static_cast<int>(item_type);
```
Consider using `base::to_underlying` here, rather than a static cast, but also check the other comment I left below, because ultimately I think no cast should be done here.

petemill replied:
We can tackle in a follow-up, this is using same methodology as the rest of the class

cdesouza-chromium commented on components/sidebar/sidebar_service.cc:
```diff
@@ -196,11 +225,41 @@ void SidebarService::MigratePrefSidebarBuiltInItemsToHidden() {
+      const auto item_id_parsed = static_cast<int>(item_type);
+      // Mark this item to be updated
+      items_are_modified = true;
+      item.SetIntKey(kSidebarItemBuiltInItemTypeKey, item_id_parsed);
```
I particularly feel this is not the best approach here. The issue of storing the key number is that now we have to make sure that the enum is stable. If I key were to be removed, that would affect the ordering, and could even lead to serious UB problems. I think this should be a string with the actual string representation for the enum key, and you should have a pair of function that converts to and from string.

petemill replied:
Yes I agree, although there are occasions chromium does similar to this too. We can discuss as a follow-up, this line is just making it compatible with existing implementation. It's not urgent to change since:
1. We do not use enum order for functional order of items (e.g. display order)
2. If we remove an item we can ignore it, or rename to something like `kHistory_OBSOLETE`. We have already removed one (for now) - `kHistory`.

cdesouza-chromium replied:
I just would like to remind the following if we go forward with this solution:

  1. The enum ordering cannot be changed as the underlying value is being recorded
  2. We cannot remove any value going forward as this will result in reordering or UB during deserialisation.
  3. Migrating away from it will lead to additional code for a migration path, which will permantely increases complexity. The code above already has migration from URL to enum key.
  4. Although we can mark keys as obsolete, that still let them in the range of valid keys, which defeats the purpose of having an enum.
 
The solution involves merely two functions.

petemill replied:
This is not going forward - the pref has been in release for some time (14 months) using this format of `static_cast<int>(item.built_in_item_type)`. It is not introduced recently.

cdesouza-chromium replied:
Ah, I see. Sorry for my misunderstanding.

cdesouza-chromium commented on components/sidebar/sidebar_service.cc:
```diff
@@ -171,12 +193,19 @@ void SidebarService::MigratePrefSidebarBuiltInItemsToHidden() {
-  for (const auto& item : items) {
+  // Find built-in items in items pref and keep them visible. Clone so that we
+  // can potentially modify and re-save.
+  auto new_items = items.Clone();
```
It would be useful to see if you can have items as a base::Value::Dict, as all the calls to items below are deprecated.

petemill replied:
done

simonhong commented on components/sidebar/sidebar_service.cc:
```diff
@@ -393,23 +473,29 @@ void SidebarService::LoadSidebarItems() {
         }
+        auto id =
+            static_cast<SidebarItem::BuiltInItemType>(*built_in_type_value);
+        auto iter = std::find_if(default_items_to_add.begin(),
```
`base::ranges::find_if`

simonhong reviewed (approved):
++

petemill commented:
I have addressed all feedback and rebased on https://github.com/brave/brave-core/pull/14800 so that it is ready. We can even merge all together in this PR for most optimized CI and merge and easy uplift.

cdesouza-chromium reviewed (approved):
LGTM
